[WIP] Onedir build

dwoz opened the pull request:
### What does this PR do?

This is an example onedir build using make to compile python and install salt. The build process does the following to make things portable:

- Modify rpath of shared libraries to look in `$ORIGIN` for other shared libraries being loaded.
- Custom shebang for installed scripts to make sure the proper python binary is used
- Customize sys.prefix and sys.exe_prefix to make sure proper python location is used
- Pin sys.path into the proper python location

### What issues does this PR fix or reference?
Fixes:

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->

Needs docs, tests changelog

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

dwoz committed:
Exmple of onedir build idea

dwoz committed:
Fix pre-commit

dwoz committed:
Optimize yum

dwoz committed:
Add install and uninstall scripts

dwoz committed:
initial deb

dwoz committed:
Fix up onedir shared libs some more

dwoz committed:
Ignore version'd onedirs

dwoz committed:
remove some cruft

dwoz committed:
Onedir build, ARM64

dwoz committed:
remove skiped package check

dwoz committed:
Fix wart

dwoz committed:
Fix wart

dwoz committed:
Always look in $ORIGIN  in addition to /libs

dwoz committed:
Initial RPM

dwoz committed:
Fix wart

dwoz committed:
test

dwoz committed:
Fix wart in fixlibs

dwoz committed:
Hacks to get tests running

dwoz committed:
meh

dwoz committed:
The fixlibs target has been removed

dwoz committed:
install libcurl

dwoz committed:
More tests

dwoz committed:
Add package dependencies to onedir build

dwoz committed:
Fix pre-commit issues

s0undt3ch commented on tests/conftest.py:
```diff
@@ -958,6 +982,12 @@ def salt_master_factory(
     config_defaults["known_hosts_file"] = tests_known_hosts_file
     config_defaults["syndic_master"] = "localhost"
     config_defaults["transport"] = salt_syndic_master_factory.config["transport"]
+    config_defaults["engines_dirs"] = [
+        str(salt_factories.get_salt_engines_path()),
+    ]
+    config_defaults["log_handlers_dirs"] = [
+        str(salt_factories.get_salt_log_handlers_path()),
+    ]
```
```suggestion
```
We should no longer require this.

s0undt3ch commented on tests/conftest.py:
```diff
@@ -676,6 +689,17 @@ def groups_collection_modifyitems(config, items):
+# ----- Fixtures Overrides ------------------------------------------------------------------------------------------>
+
+# Hook the system installation of salt into salt_factories
+@pytest.fixture(scope="session")
+def salt_factories(salt_factories, request):
+    scripts_path = request.config.getoption("--scripts-path")
+    if scripts_path:
+        salt_factories.scripts_dir = pathlib.Path(scripts_path)
+    return salt_factories
+
+
```
```suggestion
```
And instead:

```python
@pytest.fixture(scope="session")
def salt_factories_config():
    """
    Return a dictionary with the keyworkd arguments for FactoriesManager
    """
    return {
        "code_dir": str(CODE_DIR),
        "inject_coverage": MAYBE_RUN_COVERAGE,
        "inject_sitecustomize": MAYBE_RUN_COVERAGE,
        "scripts_path": request.config.getoption("--scripts-path"),
        "start_timeout": 120
        if (os.environ.get("JENKINS_URL") or os.environ.get("CI"))
        else 60,
    }
```

s0undt3ch commented on tests/pytests/conftest.py:
```diff
@@ -149,6 +149,12 @@ def salt_master_factory(
     config_defaults["reactor"] = [
         {reactor_event.tag: [str(reactor_event.sls_path)]},
     ]
+    config_defaults["engines_dirs"] = [
+        str(salt_factories.get_salt_engines_path()),
+    ]
+    config_defaults["log_handlers_dirs"] = [
+        str(salt_factories.get_salt_log_handlers_path()),
+    ]
```
```suggestion
```

s0undt3ch commented on tests/pytests/conftest.py:
```diff
@@ -308,6 +316,12 @@ def salt_minion_factory(salt_master_factory, salt_minion_id, sdb_etcd_port, vaul
         "auth": {"method": "token", "token": "testsecret", "uses": 0},
         "policies": ["testpolicy"],
     }
+    config_defaults["engines_dirs"] = [
+        str(salt_factories.get_salt_engines_path()),
+    ]
+    config_defaults["log_handlers_dirs"] = [
+        str(salt_factories.get_salt_log_handlers_path()),
+    ]
```
```suggestion
```

s0undt3ch commented on tests/pytests/conftest.py:
```diff
@@ -330,14 +344,20 @@ def salt_minion_factory(salt_master_factory, salt_minion_id, sdb_etcd_port, vaul
     config_defaults["hosts.file"] = os.path.join(RUNTIME_VARS.TMP, "hosts")
     config_defaults["aliases.file"] = os.path.join(RUNTIME_VARS.TMP, "aliases")
     config_defaults["transport"] = salt_master_factory.config["transport"]
+    config_defaults["engines_dirs"] = [
+        str(salt_factories.get_salt_engines_path()),
+    ]
+    config_defaults["log_handlers_dirs"] = [
+        str(salt_factories.get_salt_log_handlers_path()),
+    ]
```
```suggestion
```

dwoz committed:
Remove entire pycache dir instead of just pyc files

dwoz committed:
Use newest pytest-salt-factories

dwoz committed:
Fix pycache removal

s0undt3ch committed:
Updates for newer salt-factories

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Parallelize test runs

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Colors are great!

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Run against the python that the one dir ships

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Include the python headers in the onedir

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Symlink `bin/python3.X` to `bin/python` and `bin/pip3.X` to `bin/pip`

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Install `wheel` into the onedir.

This simplifies using wheel packages or building them for our
requirements.

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
More colors!

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

dwoz committed:
Temporary workaround

I wanted to avoid going down this path because it seems like we're
opening a can of worms.

Commit:
only remove the python manpages

dwoz committed:
Use as-needed to avoid libcrypt

dwoz committed:
Initial commit of MacOS onedir

dwoz committed:
Install build dependencies

Commit:
add systemd files to onedir build, this can be altered to use files included with MANIFEST.in if we move to building an sdist

Commit:
Add openssl build dependency and require that openssl to build python

dwoz committed:
Use --force-rpath for patchelf calls

dwoz committed:
migrate from fixlibs.sh to relocate.py

Commit:
Fix sed on mac

dwoz committed:
Fix warts in relocate.py

dwoz committed:
Point mac builds to proper openssl

dwoz committed:
Fix warts in relocate.py

dwoz committed:
Fix merge conflicts

dwoz committed:
fix relocate.py formatting

dwoz committed:
Move perl dependency to onedir.sh

ITJamie commented on pkg/onedir.sh:
```diff
@@ -0,0 +1,36 @@
+
+# XXX: centos 7 with dependencies installed, we need to figure out how we'll
+# handle this
+[ -z "$CONTAINER_IMAGE" ] && CONTAINER_IMAGE=dwoz1/cicd:onedir-centos7
```
Should these containers not exist in a saltstack docker hub rather than a personal docker hub account?

dwoz replied:
Yes, any docker images that are being used here will eventually be moved under the saltstack org. 

The repo for the current containers is here: https://github.com/dwoz/docker-dwoz1

dwoz committed:
fix pre-comit

Commit:
install openssl into the same location as python so relocation allocated the correct rpath

Commit:
refactor the rpath patching to its own method

Commit:
change to use correct  relative rpath

dwoz committed:
Remove Makefile

dwoz committed:
x86_64 onedir use relenv

dwoz committed:
Use pip to install relenv

dwoz committed:
Use local data dir

dwoz committed:
Use static requirements

dwoz committed:
debug environment

dwoz committed:
runtime debug

dwoz committed:
Fix typo

dwoz committed:
debug pkg install

dwoz committed:
setup python

dwoz committed:
test

dwoz committed:
meh

dwoz committed:
meh

dwoz committed:
test

dwoz committed:
Try full script path

dwoz committed:
List directory in case
